internal/discoverer/nacos.go

```diff
@@ -0,0 +1,335 @@
+			if len(entities) == 0 {
+				d.unsubscribe(cacheService)
+				delete(d.cache, serviceId)
+			}
+		}
+	}
+	return nil
+}
+
+func (d *NacosDiscoverer) Update(update *comm.Update) error {
+	values, oldArgs, newArgs, err := update.Decode()
+	if err != nil {
+		return err
+	}
+	d.defaultNamespace(&oldArgs)
+	d.defaultNamespace(&newArgs)
+
+	event, service := values[0], values[1]
+	if event != utils.EventUpdate {
+		return errors.New("incorrect update event")
+	}
+	serviceId := fmt.Sprintf("%s@%s@%s", oldArgs["namespace"], oldArgs["group"], service)
+	newServiceId := fmt.Sprintf("%s@%s@%s", newArgs["namespace"], newArgs["group"], service)
+
+	d.mu.Lock()
+	if cacheService, ok := d.cache[serviceId]; ok {
+		d.unsubscribe(cacheService)
+		d.mu.Unlock()
+
+		nodes, err := d.fetch(service, newArgs)
+		if err != nil {
+			return err
+		}
+		cacheService.nodes = nodes
+		cacheService.args = newArgs
+
+		d.mu.Lock()
+		delete(d.cache, serviceId)
+		d.cache[newServiceId] = cacheService
+		d.mu.Unlock()
+
+		watch, err := cacheService.EncodeWatch()
+		if err != nil {
+			return err
+		}
+		d.watchCh <- watch
+	}
+
+	return nil
+}
+
+func (d *NacosDiscoverer) Watch() chan *comm.Watch {
+	return d.watchCh
+}
+
+func (d *NacosDiscoverer) fetch(service string, args map[string]string) ([]Node, error) {
+	// if the namespace client has not yet been created
+	namespace := args["namespace"]
+	if _, ok := d.namingClients[namespace]; !ok {
+		clientConfig := d.ClientConfig
+		clientConfig.NamespaceId = namespace
+		client, err := clients.NewNamingClient(
+			vo.NacosClientParam{
+				ClientConfig:  &clientConfig,
+				ServerConfigs: d.ServerConfigs,
+			},
+		)
+		if err != nil {
+			return nil, err
+		}
+		d.namingClients[namespace] = client
+	}
+
+	client := d.namingClients[namespace]
+	serviceInfo, err := client.GetService(vo.GetServiceParam{
+		ServiceName: service,
+		GroupName:   args["group"],
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// watch the new service
+	if err := d.subscribe(service, args); err != nil {
+		return nil, err
+	}
+
+	nodes := make([]Node, len(serviceInfo.Hosts))
+	for i, host := range serviceInfo.Hosts {
+		address := fmt.Sprintf("%s:%d", host.Ip, host.Port)
+		weight := int(host.Weight)
+		if weight == 0 {
+			weight = d.weight
+		}
+
+		nodes[i] = Node{
+			host:   address,
+			weight: weight,
+		}
+	}
+
+	return nodes, nil
+}
+
+func (d *NacosDiscoverer) watch(serviceId string) func([]model.SubscribeService, error) {
+	watch := func(services []model.SubscribeService, err error) {
+		nodes := make([]Node, len(services))
+		for i, inst := range services {
+			address := fmt.Sprintf("%s:%d", inst.Ip, inst.Port)
+			weight := int(inst.Weight)
+			if weight == 0 {
+				weight = d.weight
+			}
+
+			nodes[i] = Node{
+				host:   address,
+				weight: weight,
+			}
+		}
+
+		d.mu.Lock()
+		cacheService := d.cache[serviceId]
+		cacheService.nodes = nodes
+		d.cache[serviceId] = cacheService
```

Comment:
We don't need this line?

Reply:
This line was required in the old code, but in the new implementation `cache` will store the reference, so this line can be omitted.